src/chain/electrum.rs

```diff
@@ -0,0 +1,303 @@
+					log_trace!(
+						self.logger,
+						"Failed broadcast transaction bytes: {}",
+						log_bytes!(tx_bytes)
+					);
+				},
+			},
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to broadcast transaction due to timeout {}: {}",
+					txid,
+					e
+				);
+				log_trace!(
+					self.logger,
+					"Failed broadcast transaction bytes: {}",
+					log_bytes!(tx_bytes)
+				);
+			},
+		}
+	}
+
+	pub(crate) async fn get_fee_rate_cache_update(
+		&self,
+	) -> Result<HashMap<ConfirmationTarget, FeeRate>, Error> {
+		let electrum_client = Arc::clone(&self.electrum_client);
+
+		let mut batch = Batch::default();
+		let confirmation_targets = get_all_conf_targets();
+		for target in confirmation_targets {
+			let num_blocks = get_num_block_defaults_for_target(target);
+			batch.estimate_fee(num_blocks);
+		}
+
+		let spawn_fut = self.runtime.spawn_blocking(move || electrum_client.batch_call(&batch));
+
+		let timeout_fut = tokio::time::timeout(
+			Duration::from_secs(FEE_RATE_CACHE_UPDATE_TIMEOUT_SECS),
+			spawn_fut,
+		);
+
+		let raw_estimates_btc_kvb = timeout_fut
+			.await
+			.map_err(|e| {
+				log_error!(self.logger, "Updating fee rate estimates timed out: {}", e);
+				Error::FeerateEstimationUpdateTimeout
+			})?
+			.map_err(|e| {
+				log_error!(self.logger, "Failed to retrieve fee rate estimates: {}", e);
+				Error::FeerateEstimationUpdateFailed
+			})?
+			.map_err(|e| {
+				log_error!(self.logger, "Failed to retrieve fee rate estimates: {}", e);
+				Error::FeerateEstimationUpdateFailed
+			})?;
+
+		if raw_estimates_btc_kvb.len() != confirmation_targets.len()
+			&& self.config.network == Network::Bitcoin
+		{
+			// Ensure we fail if we didn't receive all estimates.
+			debug_assert!(false,
+				"Electrum server didn't return all expected results. This is disallowed on Mainnet."
+			);
+			log_error!(self.logger,
+				"Failed to retrieve fee rate estimates: Electrum server didn't return all expected results. This is disallowed on Mainnet."
+			);
+			return Err(Error::FeerateEstimationUpdateFailed);
+		}
+
+		let mut new_fee_rate_cache = HashMap::with_capacity(10);
+		for (target, raw_fee_rate_btc_per_kvb) in
+			confirmation_targets.into_iter().zip(raw_estimates_btc_kvb.into_iter())
+		{
+			// Parse the retrieved serde_json::Value and fall back to 1 sat/vb (10^3 / 10^8 = 10^-5
+			// = 0.00001 btc/kvb) if we fail or it yields less than that. This is mostly necessary
+			// to continue on `signet`/`regtest` where we might not get estimates (or bogus
+			// values).
+			let fee_rate_btc_per_kvb = raw_fee_rate_btc_per_kvb
+				.as_f64()
+				.map_or(0.00001, |converted| converted.max(0.00001));
+
+			// Electrum, just like Bitcoin Core, gives us a feerate in BTC/KvB.
+			// Thus, we multiply by 25_000_000 (10^8 / 4) to get satoshis/kwu.
+			let fee_rate = {
+				let fee_rate_sat_per_kwu = (fee_rate_btc_per_kvb * 25_000_000.0).round() as u64;
```

Comment:
No risk here in rounding (down)?

Reply:
Ah yes it's kwu. I was worried about somehow rounding down to zero. But I suppose with that precision it doesn't matter which way to round?